### PR TITLE
wlr drag icon

### LINKED ISSUE #188: Implement wlr_drag_icon
The `wlr_drag_icon` represents a drag icon that should be rendered on the screen at the cursor or touch location during a drag-and-drop operation.

Drag icons appear on the `wlr_seat` `new_drag_icon` event in an unmapped state.

*Drag icon events*

* `map` - the drag icon is ready to be rendered on the output
* `destroy` - the drag icon is being destroyed

*Drag icon surface events*

* `commit` - the drag icon should be damaged on the output

*Rendering*

Drag icons should be stored in a list on the seat. Drag icons should only be rendered if the `mapped` bit is set on the drag icon struct. Drag icons have `sx` and `sy` members which represent the offset from the cursor position the drag icon should be rendered at. These should be added to the layout coordinates of the cursor (or touch point) position whenever the cursor position changes and rendered at that location during the rendering loop.

Drag icons (as all `wlr_surface`s) are capable of having subsurfaces, but we've never seen a drag icon implementation that uses this nor do we expect this to be a use case in the future.

### PR DESCRIPTION
**Pull request by acrisci**
fixes #188

### CODE REVIEW

**Comment by Timidger** [comment]
Please rebase against master

File: src/types/seat/seat.rs [hunk 1]
```diff
@@ -256,9 +256,24 @@ wayland_listener!(Seat, (*mut wlr_seat, Box<SeatHandler>), [
...
+            let surface_state = (*(*data).surface).data as *mut InternalSurfaceState;
+            (*(*surface_state).surface).data().1 = surface_handler;
+        }
+
+        if let Some(drag_icon_handler) = drag_icon_handler {
+            let mut listener = DragIconListener::new((drag_icon, drag_icon_handler));
+            wl_signal_add(&mut (*data).events.destroy as *mut _ as _,
+                          listener.destroy_listener() as _);
+            wl_signal_add(&mut (*data).events.map as *mut _ as _,
+                          listener.map_listener() as _);
```
Comment (acrisci) [inline src/types/seat/seat.rs]: How do I keep it from dropping?

Comment (Timidger) [reply src/types/seat/seat.rs]: `::std::mem::forget` [docs here](https://doc.rust-lang.org/beta/std/mem/fn.forget.html)

**Comment by acrisci** [comment]
There's a bit of churn in the wlroots drag icon interface in [wlroots#1043](https://github.com/swaywm/wlroots/pull/1043)

**Comment by Timidger** [comment]
Got it no worries. Let's wait for that to shake out then.

File: src/types/seat/seat.rs (hunk 1 above)
Comment (acrisci) [reply src/types/seat/seat.rs]: Oh that's a pretty cool pattern. Why don't we do that everywhere?

**Comment by Timidger** [comment]
Is this ready to be reviewed?

**Comment by acrisci** [comment]
Yeah, there's some warnings for unused things that I'll remove after the first review.

File: src/types/seat/seat.rs [hunk 2]
```diff
@@ -237,6 +244,40 @@ wayland_listener!(Seat, (*mut wlr_seat, Box<SeatHandler>), [
...
     };
+    new_drag_icon_listener => new_drag_icon_notify: |this: &mut Seat, data: *mut libc::c_void,|
+    unsafe {
+        let (seat_ptr, ref mut handler) = this.data;
+        let data = data as *mut wlr_drag_icon;
+        let compositor = match compositor_handle() {
+            Some(handle) => handle,
+            None => return
+        };
+        let seat = Seat::from_ptr(seat_ptr);
```
Comment (Timidger) [inline src/types/seat/seat.rs]: Use `SeatHandle::from_ptr` here so you don't have to `Box::into_raw` it later

File: src/types/seat/drag_icon.rs [hunk 3]
```diff
@@ -0,0 +1,162 @@
...
+            let surface = (*self.drag_icon).surface;
+			if surface.is_null() {
+				panic!("drag icon had a null surface!");
+			}
+			SurfaceHandle::from_ptr(surface)
+        }
+    }
+
+    /// The position to place the surface of the drag icon relative to the cursor position
+	pub fn position(&mut self) -> (i32, i32) {
```
Comment (Timidger) [inline src/types/seat/drag_icon.rs]: Indentation

File: src/types/seat/drag_icon.rs [hunk 4]
```diff
@@ -0,0 +1,162 @@
...
+            Err(err) => panic::resume_unwind(err)
+        }
+    }
+
+    unsafe fn as_ptr(&self) -> *mut wlr_drag_icon {
+        self.drag_icon
+    }
+}
+
+pub(crate) struct DragIconState {
```
Comment (Timidger) [inline src/types/seat/drag_icon.rs]: Put struct definitions at top of file

File: src/types/seat/seat.rs [hunk 5]
```diff
@@ -237,6 +244,40 @@ wayland_listener!(Seat, (*mut wlr_seat, Box<SeatHandler>), [
...
+            wl_signal_add(&mut (*data).events.destroy as *mut _ as _,
+                          listener.destroy_listener() as _);
+            wl_signal_add(&mut (*data).events.map as *mut _ as _,
+                          listener.map_listener() as _);
+            wl_signal_add(&mut (*data).events.unmap as *mut _ as _,
+                          listener.unmap_listener() as _);
+            Box::into_raw(listener);
+        }
+
+        println!("New drag icon request {:p}", data);
```
Comment (Timidger) [inline src/types/seat/seat.rs]: Use `wlr_log!(L_DEBUG, "New drag icon request {:p}", data);`

File: src/types/seat/drag_icon.rs [hunk 6]
```diff
@@ -0,0 +1,162 @@
...
+            liveliness,
+            drag_icon
+        }
+    }
+
+    /// Get a handle for the surface associated with this drag icon
+    pub fn surface(&mut self) -> SurfaceHandle {
+        unsafe {
+            let surface = (*self.drag_icon).surface;
+			if surface.is_null() {
```
Comment (Timidger) [inline src/types/seat/drag_icon.rs]: More weird indentation?

File: src/types/seat/drag_icon.rs [hunk 7]
```diff
@@ -0,0 +1,162 @@
...
+				panic!("drag icon had a null surface!");
+			}
+			SurfaceHandle::from_ptr(surface)
+        }
+    }
+
+    /// The position to place the surface of the drag icon relative to the cursor position
```
Comment (Timidger) [inline src/types/seat/drag_icon.rs]: Specify it's in `(sx, sy)` format like the other doc comments.

File: src/types/seat/drag_icon.rs (hunk 6 above)
Comment (acrisci) [reply src/types/seat/drag_icon.rs]: My editor is messing up and doing tabs sometimes for some reason.

**Comment by Timidger** [comment]
Thanks!

**Pull request merged**